enclave/crypto/random.c

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "random.h"
+#include <mbedtls/entropy.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/random.h>
+#include <openenclave/internal/thread.h>
+
+/*
+**==============================================================================
+**
+** Local definitions
+**
+**==============================================================================
+*/
+
+static mbedtls_ctr_drbg_context _drbg;
+static mbedtls_entropy_context _entropy;
+
+static oe_result_t _seed_entropy_source()
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    mbedtls_ctr_drbg_init(&_drbg);
+    mbedtls_entropy_init(&_entropy);
+
+    OE_CHECK((oe_result_t)mbedtls_ctr_drbg_seed(
+        &_drbg, mbedtls_entropy_func, &_entropy, NULL, 0));
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+static oe_result_t _seed_result = OE_UNEXPECTED;
+static oe_once_t _seed_once = OE_ONCE_INIT;
+
+/* Wrapper to set file-scope _seed_result */
+static void _seed_entropy_source_once()
+{
+    _seed_result = _seed_entropy_source();
+}
+
+mbedtls_ctr_drbg_context* oe_mbedtls_get_drbg()
```

Comment:
This function does not check the seed result. Is that by design?

Reply:
That's how the function was implemented. @mikbras do you know why the seed result isn't checked.

Reply:
Regardless of the intent, this looks like a bug; the expectation is that this function should return NULL on failure (see enclave/ec.c or rsa.c), but `&_drbg` is always a valid pointer, regardless of whether it was ever initialized or successfully seeded.

Reply:
Ok. Sounds good. It looked like a bug to me too, so I wasn't sure why it was implemented that way.